eclipse-collections/src/main/java/org/eclipse/collections/impl/map/immutable/ImmutableUnifiedMap.java

```diff
@@ -120,9 +118,9 @@ public void forEachKeyValue(Procedure2<? super K, ? super V> procedure)
     }
 
     @Override
-    public Set<K> keySet()
+    public ImmutableKeySet<K> keySet()
     {
-        return UnmodifiableMutableSet.of(this.delegate.keySet());
+        return new ImmutableKeySet<>(this.delegate.keySet().toImmutable());
```

Comment:
I wonder if this can be done by returning an ImmutableWrapper around the mutable keySet so it doesn't have to be copied.